drivers: flash: spi_nor: support devices that default to protect blocks

pabigot opened the pull request:
Some SPI NOR devices, particularly Atmel and SST, power-up with block
protect bits set in the status register.  These bits must be cleared
before any erase or program operation can succeed.  However, blindly
clearing bits in SR is wrong as some of these are non-volatile and
control chip behavior, including quad-enable.

Add a devicetree flag to identify device-specific BP bits in the status
register that should be cleared on startup only for devices that need
them, and when set do the clear during initialization.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: flash: spi_nor: support devices that default to protect blocks

Some SPI NOR devices, particularly Atmel and SST, power-up with block
protect bits set in the status register.  These bits must be cleared
before any erase or program operation can succeed.  However, blindly
clearing bits in SR is wrong as some of these are non-volatile and
control chip behavior, including quad-enable.

Add a devicetree flag to identify device-specific BP bits in the status
register that should be cleared on startup only for devices that need
them, and when set do the clear during initialization.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
This is an alternative implementation to solve the problem that #32548 addresses in a more generic and safer way.

Note: has-lock derives from the SPI_NOR_HAS_LOCK flag that's associated with devices in the Linux device registries.

pabigot commented:
See also https://github.com/zephyrproject-rtos/zephyr/pull/32548/files#r583099791

Note that I don't have any devices with this requirement, so I can't verify that it works.  Need input from @shlomow and @ycsin.

ycsin commented:
> See also https://github.com/zephyrproject-rtos/zephyr/pull/32548/files#r583099791
> 
> Note that I don't have any devices with this requirement, so I can't verify that it works. Need input from @shlomow and @ycsin.

I think it works, but as I'm using PlatformIO and thus not able to use the latest Zephyr ver, I'm not able to copy and paste everything. Intead I added the following lines:

``` C
	/* Check for block protect bits that need to be cleared. */
	if (1) {
		rc = spi_nor_rdsr(dev);

		/* Only clear if RDSR worked and something's set. */
		if (rc > 0) {
			rc = spi_nor_wrsr(dev, rc & ~(0X1));
		}

		if (rc != 0) {
			LOG_ERR("BP clear failed: %d\n", rc);
			return -ENODEV;
		}
	}
```

And the `spi_nor_rdsr` & `spi_nor_wrsr` function, it seems to work with `samples/subsys/settings`. But I have a question: are `spi_nor_cmd_write(dev, SPI_NOR_CMD_WREN);` and `spi_nor_wait_until_ready(dev);` not required for this operation?

pabigot commented:
> I think it works, but as I'm using PlatformIO and thus not able to use the latest Zephyr ver, I'm not able to copy and paste everything (bindings related). Instead I simply added the following lines:
> 			rc = spi_nor_wrsr(dev, rc & ~(0X1));

I think the mask for the BP bits of SR should be 0x7C (BP4..BP0) for the AT25SF128A, since the bit masked by 0x01 is the write-in-progress status byte (which isn't writable).  But check your datasheet.

> And the `spi_nor_rdsr` & `spi_nor_wrsr` function, it seems to work with `samples/subsys/settings`. But I have a question: are `spi_nor_cmd_write(dev, SPI_NOR_CMD_WREN);` and `spi_nor_wait_until_ready(dev);` not required for this operation?

They shouldn't be, since the command doesn't write to the memory, and there are no earlier operations that could still be in progress.

ycsin commented:
> > I think it works, but as I'm using PlatformIO and thus not able to use the latest Zephyr ver, I'm not able to copy and paste everything (bindings related). Instead I simply added the following lines:
> > rc = spi_nor_wrsr(dev, rc & ~(0X1));
> 
> I think the mask for the BP bits of SR should be 0x7C (BP4..BP0) for the AT25SF128A, since the bit masked by 0x01 is the write-in-progress status byte (which isn't writable). But check your datasheet.
> 
> > And the `spi_nor_rdsr` & `spi_nor_wrsr` function, it seems to work with `samples/subsys/settings`. But I have a question: are `spi_nor_cmd_write(dev, SPI_NOR_CMD_WREN);` and `spi_nor_wait_until_ready(dev);` not required for this operation?
> 
> They shouldn't be, since the command doesn't write to the memory, and there are no earlier operations that could still be in progress.

Yes you are right, from AT25DF081A datasheet:
![image](https://user-images.githubusercontent.com/1563974/109253764-706c3a80-782b-11eb-959a-d06962f8a8c3.png)
![image](https://user-images.githubusercontent.com/1563974/109253393-a957df80-782a-11eb-96b1-0e76eb1d32bc.png)
![image](https://user-images.githubusercontent.com/1563974/109253415-b83e9200-782a-11eb-9028-9ecbb1e4ac14.png)

I will try again later.
Also is it not necessary to reprotect the chip after erasing as well as unprotect sector prior to writing and reprotect sector after that?

pabigot commented:
> Also is it not necessary to reprotect the chip after erasing as well as unprotect sector prior to writing and reprotect sector after that?

Only if your application requires extra protection (in which case you probably shouldn't be removing BP on startup, but instead doing that before erase/write and restoring it after erase/write).  But there is no support for write-protecting regions in the current Zephyr flash API, so you'd have to do that another way.

ycsin commented:
Hey @shlomow , I've tried this pr but found that it is necessary to have

```
spi_nor_cmd_write(dev, SPI_NOR_CMD_WREN);
```

before `spi_nor.c:930` for the erase and write command to work. Having

```
spi_nor_wait_until_ready(dev);
```

after `spi_nor.c:930` or not doesn't affect anything for me, at least for the `samples/flash/spi_flash` example.

But I only tested this on Zephyr v2.4.0 as I'm having some problem with the latest branch, could you please verify?

pabigot commented:
>I've tried this pr but found that it is necessary to have [WREN]

You're correct; I didn't read the WRSR specification closely enough.  PR has been updated to issue WREN before WRSR.

shlomow commented:
@ycsin I checked the datasheet again and I think you are right that there is no need to call `spi_nor_wait_until_ready`. By the way I looked at a non legacy device datasheet (micron n25q) and there is a need to call this function after writing the status register. I will try to test this tomorrow.

@pabigot I think that the `has-lock` binding can be a little bit confusing since it also stands for non legacy devices and is supported in jedec spec. If in the future there will be `lock / unlock` api for flash devices in zephyr than this will need some refactoring.

pabigot commented:
@shlomow what do you mean by "legacy device"?

I don't see anything in JESD216 that relates to `has-lock`; the only reference to locking that shows up is in DWORD 1 indicating support for individual sector write lock instructions.

I've added a fix that makes sure the device is ready before issuing commands, but on further consideration I think I'm going to rework that a little more.

shlomow commented:
@pabigot By legacy I mean a device that wakes up with all it sectors protected.
I tested this now, and it worked for me

pabigot commented:
I'd like to get this in soon as I have to make further touches to this driver that I'll base on this.

pabigot commented:
@nvlsianpu Please review and get this merged so there isn't another change to this file that slips in and adds conflicts like #33216 did.

This blocks #32844 which is starting to become more important.

carlescufi reviewed (approved):
LGTM

jfischer-no reviewed (approved):
looks reasonable

nvlsianpu reviewed (commented):
[Performance]: Each write/erase operation normally returns once spi_nor_wait_until_ready() returns as ready. Additional check gives us more latency. Maybe additional check might be done only for the firs call?

de-nordic commented on drivers/flash/spi_nor.c:
```diff
@@ -307,10 +310,34 @@ static int spi_nor_access(const struct device *const dev,
+	uint8_t reg;
+
+	do {
+		ret = spi_nor_cmd_read(dev, SPI_NOR_CMD_RDSR, &reg, 1);
```
Use `sizeof(reg)` instead of 1, the same way you did in `spi_nor_rdsr`

de-nordic reviewed (commented):
IMHO `acquire_device` should not be concluded with the `spi_nor_wait_until_ready` and the `spi_nor_wait_until_ready` should be only called before operation, as it causes spi transfers (I might be wrong here), that requires the device; for example lets look at current write loop in `spi_nor_write`, starting at line 544 just before loop:

*544*: `acquire_device()` -- will wait at the end with `spi_nor_wait_until_ready`, not really needed we might be looping on spi for longer than needed if next operation on device  is time distant from return from `acquire_device`;
*545*: ` spi_nor_write_protection_set` -- will begin with `spi_nor_wait_until_ready`, this is OK because it requires the device to be ready to operate on it; it also makes the wait in `acquire_device()` redundant'

Next check should happen in loop, just before line *561* because here is where we expect the device to be ready and all operations that happen before will add to the required time; the `spi_nor_wait_until_ready` in line *572* would not be needed if we have before *561* and we are also covered when the loop ends because the ` spi_nor_write_protection_set` in line *576*, that sets the protection back will call `spi_nor_wait_until_ready` on entry.

Additional example in `spi_nor_configure`, where `acquire_device` (line *961*) is followed by `rc = spi_nor_rdsr(dev);`  which means that we have been just reading the status register in loop (`acquire_device->spi_nor_wait_until_ready`) to read it again to check different bits.

pabigot commented:
@de-nordic thanks, good catch.

When PR was written the driver waited for ready in:

* (B) spi_nor_read before nor_cmd_read
* (A) spi_nor_write after each nor_cmd_write
* (A) spi_nor_erase after each nor_cmd erase
* (B) spi_nor_write_protection_set before first write (but not second)
* (B) spi_nor_sfdp_read before read
* (B) spi_nor_read_jedec_id before read

(#33216 changed some of that, as there were a lot of collisions.)

These showed an inconsistent rule for when the busy check is required: does the driver return before the operation is complete at the device level (so the checks tagged `B` above are required), or does it return after completion (via the checks tagged `A` above)?  If the former there are a lot of waits missing; if the latter there's no reason to check before any operation, nor after acquire.

So originally I thought it was necessary to wait before operations.  Now I think it's more correct to wait after: yes we lose a little bit of time that we could theoretically overlap with things that don't try to send commands, but if we waited we'd have a *lot* more places to insert the check before sending another command.

The Linux spi_nor driver also consistently waits for the operation to complete after WRITE_STATUS, ERASE, and PROGRAM, and nowhere else.

So I've reworked this to do exactly that, and document that usage.  That leaves three places in the code where the wait is required.

de-nordic reviewed (approved):
Looks OK.
Thanks for the improvements.

nvlsianpu commented:
@pabigot looks fine now. Can you correct the firs commit message?

pabigot committed:
drivers: flash: spi_nor: clean up ready detection

Reads would wait until the device was ready before issuing the first
command; writes and erases did not.  Fix this documenting and changing
so that wait-for-ready is invoked only where needed, i.e. to confirm
that a WRITE_STATUS, ERASE, or PROGRAM operation has completed before
proceeding to allow more commands to be submitted.  This matches Linux
spi_nor driver behavior.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: flash: spi_nor: document caller requirements

Add the standard note to two functions that require the caller to have
first acquired the device.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
> @pabigot looks fine now. Can you correct the firs commit message?

Done: I needed to split it, since it not only changed the policy for ready wait it also had two documentation updates.  PR content has not changed, just the commits and messages.
